src/transformers/modeling_longformer.py

```diff
@@ -0,0 +1,667 @@
+        masked_lm_loss (`optional`, returned when ``masked_lm_labels`` is provided) ``torch.FloatTensor`` of shape ``(1,)``:
+            Masked language modeling loss.
+        prediction_scores (:obj:`torch.FloatTensor` of shape :obj:`(batch_size, sequence_length, config.vocab_size)`)
+            Prediction scores of the language modeling head (scores for each vocabulary token before SoftMax).
+        hidden_states (:obj:`tuple(torch.FloatTensor)`, `optional`, returned when ``config.output_hidden_states=True``):
+            Tuple of :obj:`torch.FloatTensor` (one for the output of the embeddings + one for the output of each layer)
+            of shape :obj:`(batch_size, sequence_length, hidden_size)`.
+
+            Hidden-states of the model at the output of each layer plus the initial embedding outputs.
+        attentions (:obj:`tuple(torch.FloatTensor)`, `optional`, returned when ``config.output_attentions=True``):
+            Tuple of :obj:`torch.FloatTensor` (one for each layer) of shape
+            :obj:`(batch_size, num_heads, sequence_length, sequence_length)`.
+
+            Attentions weights after the attention softmax, used to compute the weighted average in the self-attention
+            heads.
+
+    Examples::
+
+        import torch
+        from transformers import LongformerModel, LongformerTokenizer
+
+        model = LongformerModel.from_pretrained('longformer-base-4096')
+        tokenizer = LongformerTokenizer.from_pretrained('longformer-base-4096')
+
+        SAMPLE_TEXT = ' '.join(['Hello world! '] * 1000)  # long input document
+        input_ids = torch.tensor(tokenizer.encode(SAMPLE_TEXT)).unsqueeze(0)  # batch of size 1
+
+        # Attention mask values -- 0: no attention, 1: local attention, 2: global attention
+        attention_mask = torch.ones(input_ids.shape, dtype=torch.long, device=input_ids.device) # initialize to local attention
+        attention_mask[:, [1, 4, 21,]] = 2  # Set global attention based on the task. For example,
+                                            # classification: the <s> token
+                                            # QA: question tokens
+        sequence_output, pooled_output = model(input_ids, attention_mask=attention_mask)
+        """
+
+        # padding
+        attention_window =  \
+            self.config.attention_window if isinstance(self.config.attention_window, int) else max(self.config.attention_window)
+        padding_len, input_ids, attention_mask, token_type_ids, position_ids, inputs_embeds = self._pad_to_window_size(
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            token_type_ids=token_type_ids,
+            position_ids=position_ids,
+            inputs_embeds=inputs_embeds,
+            attention_window=attention_window,
+            pad_token_id=self.config.pad_token_id)
+
+        # embed
+        output = super().forward(
```

Comment:
I guess I would prefer to copy past the `forward` function of Roberta into the model here, but I guess ok for now! What do you think @thomwolf @LysandreJik ?

Reply:
I am curious why you think this is better. I was assuming it is the other way around, that less duplicate code is better.

Reply:
I guess that's actually a very fundamental point of implementing Longformer so here is my opinion: 

For me, `RobertaModel` == `BertModel` because it's literally the exact the same architecture except for the word embeddings which are completely overwritten in `RobertaModel` - so it's ok for me that `RobertaModel` inherits the complete forward function from Bert. `Longformer` on the other hand uses a lot of the things from `RobertaModel`: Feed Forward layers, the word embeddings (in it's expanded version), ..., but is essentially a different architecture and needs a slightly different forward function. It feels a bit to me that we try to wrap certain Longformer functionality around Roberta functionality and when it doesn't work overwrite very specific parts of `Roberta` logic with `Longformer` logic (longformer attention layer), whereas for me `Longformer` is essentially a different model. 

For me the downsides of using Roberta/Bert code as much as possible, but having a completely different attention layers are the following:

1. _Flexibility to add code / change code_: 
a) The attention layers are usually the most important part of a model so that functions at model level insert parameters for the attention layer. This being said, adding new features for Longformer won't really be possible without changing Bert code. Let's say we decide to also add a auto-regressive longformer version `LonformerWithLMHead` (not masked Longformer model) and want to add a caching mechanism to store past key-value states as is done in `gpt2`, `t5` ... (and soon `reformer`), we will anyways need a separate `forward` function. 

b) Longformer forward is completely dependant on Bert forward:
Every change to Bert's forward function affects Longformer (and Roberta). For Roberta, I think this is alright because it's the exact same architecture (same attention layer), so a change in `attention_masks`, `encoder_hidden_states`, etc. are fine, but could lead to problems in Longformer because it has a different attention layer. I'm a bit scared that it might make us write the Longformer code in a way that would fit Bert's forward function, but doesn't show the logic of longformer anymore in a clear way

2. _Readability_:
If a user wants to track what happens in Longformer when calling model(input_ids), he will have to jump from longformer's part forward to Roberta's non-existing forward to finally Bert. Errors will be very hard to track and understand.

The only upside for me is less duplicated code, but I very much prefer readability over duplicated code :-)

IMO, to keep maximum flexibility for Longformer I would go as far as to add a `LongformerLayer` class with its own `forward` function **and** a `LongformerAttention` class with its own forward function. I would then fill the layers in those classes either with "Bert components" (like `BertOutput`, `BertIntermediate`, ...) or the `LongformerSelfAttention`. Then we would not have to do this very specific layer injection at `LongformerModel` level. 

Also, people will soon ask about a encoder-decoder version for `Longformer` and I think at this point we probably should (or have to) rewrite the class to be able to handle cross longformer attention layers and longformer self attention layers well. I think the similarity between Bert/Roberta and Longformer is not big enough for complete model inheritance. 

Sorry for that long text :D 
But I guess I'm also quite biased, so very keen to discuss this :-) 


@LysandreJik, @thomwolf, @sshleifer  - what do you think?

Reply:
I agree with @patrickvonplaten.

One of the specificities of `transformers` (quite different from other libraries indeed) is to have each model code almost fully contained in a single file. This allows in particular people to easily extend the models.

We should do the same here and have the full code of Bert copied and modified here (at the expense of code duplication but it's fine for us), I think.

Reply:
I sort of agree with @ibeltagy 
 
(1) I think the PR that makes `LongformerEncoderDecoder` should copy paste if needed, but it is not required in this PR. I would also imagine that that PR will use this as the encoder and then write a new `LongformerDecoder` whose `forward` method supports incremental decoding.

(2) Most of the papers I've read describe their contributions on top of Bert, so it's nice to have the code reflect that abstraction.

(3) I think this is extremely readable. It's very clear that an external `forward` method is being called.

Reply:
I second the arguments of @sshleifer. 

Here's a different suggestion for how to implement Longformer; only keep `LongformerSelfAttention` (no `LongformerModel`) and add an if statement inside various models to switch between `BertSelfAttention` and `LongformerSelfAttention` using a configuration. 
Pros: can convert existing models into their long version
Cons: a large code change
I don't think we should do this now though, just keep it in mind for future improvements.

Reply:
Hmm, don't really agree with the arguments @sshleifer :D 

1) - I don't really think we should  add much code for the Encoder-Decoder version besides the `cross_attention_layer`. We should use the EncoderDecoderModel from https://github.com/huggingface/transformers/blob/7defc6670fa76e857109e1b99f3e919da8d11f42/src/transformers/modeling_encoder_decoder.py#L29 to create the encoder and decoder from  `LongformerModel` as it's already done in Bert. There should not be a `LongformerDecoder` IMO.

2) - I think the code should show which `Roberta/Bert` components were used and which components are new. I don't think we need a high level abstraction for this.

3) - Most people have installed the library with pip. If they run `Longformer.forward` and see `Bert` specific error messages, I could see this leading to confusion. 

I agree though that this (big) change does not have to be done in this PR though, but can be done at a later stage.

Reply:
Will keep this to a separate PR.